src/Core/Caching/InMemoryCacheClient.cs

```diff
@@ -203,9 +205,84 @@ public class InMemoryCacheClient : MaintenanceBase, ICacheClient {
             if (expiresIn.HasValue)
                 ScheduleNextMaintenance(expiresAt);
 
+            Cleanup();
+
             return difference;
         }
 
+        public async Task<bool> SetAddAsync<T>(string key, T value, TimeSpan? expiresIn = null) {
+            // TODO: Look up the existing expiration if expiresIn is null.
+            DateTime expiresAt = expiresIn.HasValue ? DateTime.UtcNow.Add(expiresIn.Value) : DateTime.MaxValue;
+            if (expiresAt < DateTime.UtcNow) {
+                _logger.Trace("Removing expired key {0}", key);
+
+                await this.RemoveAsync(key).AnyContext();
+                await OnItemExpiredAsync(key).AnyContext();
+                return false;
+            }
+
+
+            var entry = new CacheEntry(new List<T> { value }, expiresAt, ShouldCloneValues);
+            _memory.AddOrUpdate(key, entry, (k, cacheEntry) =>
+            {
+                var collection = cacheEntry.Value as ICollection<T>;
+                collection?.Add(value);
+                cacheEntry.Value = collection;
```

Comment:
I guess if someone else set a normal value in this key it might be a problem. Maybe we should bail throw or just bail and return false if it's not a collection.